src/backtrace.c

```diff
@@ -0,0 +1,720 @@
+ *    for an instruction saving the RA register to the stack (eg: `sd $ra, nn($sp)`), and an instruction
+ *    creating the stack frame (eg: `addiu $sp, $sp, -nn`). Once both are found, the heuristic knows how to
+ *    fill in `.stack_size` and `.ra_offset` fields of the function description structure, and it can stop.
+ *  * Some functions also modify $fp (the frame pointer register): sometimes, they just use it as one additional
+ *    free register, and other times they really use it as frame pointer. If the heuristic finds the
+ *    instruction `move $fp, $sp`, it knows that the function uses $fp as frame pointer, and will mark
+ *    the function as BT_FUNCTION_FRAMEPOINTER. In any case, the field `.fp_offset` will be filled in
+ *    with the offset in the stack where $fp is stored, so that the backtrace engine can track the
+ *    current value of the register in any case.
+ *  * The 0.01% of the functions that do not have a stack frame but appear in the call stack are leaf
+ *    functions interrupted by exceptions. Leaf functions pose two important problems: first, $ra is
+ *    not saved into the stack so there is no way to know where to go back. Second, there is no clear
+ *    indication where the function begins (as we normally stops analysis when the see the stack frame
+ *    creation). So in this case the heuristic would fail. We rely thus on two hints coming from the caller:
+ *    * First, we expect the caller to set from_exception=true, so that we know that we might potentially
+ *      deal with a leaf function. 
+ *    * Second, the caller should provide the function start address, so that we stop the analysis when
+ *      we reach it, and mark the function as BT_LEAF.
+ *    * If the function start address is not provided (because e.g. the symbol table was not found and
+ *      thus we have no information about function starts), the last ditch heuristic is to look for
+ *      the nops that are normally used to align the function start to the FUNCTION_ALIGNMENT boundary.
+ *      Obviously this is a very fragile heuristic (it will fail if the function required no nops to be
+ *      properly aligned), but it is the best we can do. Worst case, in this specific case of a leaf
+ *      function interrupted by the exception, the stack trace will be wrong from this point on.
+ * 
+ * @param func                        Output function description structure
+ * @param ptr                         Pointer to the function code at the point where the backtrace starts.
+ *                                    This is normally the point where a JAL opcode is found, as we are walking
+ *                                    up the call stack.
+ * @param func_start                  Start of the function being analyzed. This is optional: the heuristic can work
+ *                                    without this hint, but it is useful in certain situations (eg: to better
```

Comment:
It is actually not used anywhere else for now, but considering it may be used in the future and it is not the same thing as `from_exception` I think it is ok to have it as a separate parameter. Just thinking out loud.

Reply:
Yeah I thought the same, for now I'd just leave it separate